feat: Custom bank quick jump pages

ShadowCat117 opened the pull request:
https://github.com/Wynntils/Artemis/assets/8337467/baa4a802-dc96-4122-a6dd-248ece3067af

Once this is reviewed and cleaned up I'll add it for the misc bucket as a seperate feature

ShadowCat117 committed:
feat: Custom bank quick jump pages

ShadowCat117 committed:
Merge branch 'main' into CustomBankQuickJumpButtons

Commit:
ci: spotless formatting

ryanzhoudev commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,221 @@
+
+        StyledText title = StyledText.fromComponent(screen.getTitle());
+
+        Matcher pageMatcher = title.getMatcher(BANK_PAGE_MATCHER);
```
Can you use ContainerModel#isBankScreen instead? If you need access to the page numbers, you should also implement a helper method there.

ryanzhoudev commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,221 @@
+            }
+        } else {
+            return;
+        }
```
This entire else clause is not needed. Also, for consistency, when you switch to the ContainerModel method, you can just do an inverted check like you did at the top.

kristofbolyai replied:
Marked as resolved but you didnt change it?

ShadowCat117 replied:
I think you're viewing an old version

Fixed in latest https://github.com/Wynntils/Artemis/blob/8a79c17154af466804098f740b263269a326fd7a/common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java

ryanzhoudev commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,221 @@
+
+@ConfigCategory(Category.INVENTORY)
+public class CustomBankPagesFeature extends Feature {
+    // If possible, lock these to only allow between 1-21
```
You could use the onConfigUpdate event to check if any of these are invalid, then reset it and tell the user to fix it.

magicus replied:
Or we should extend our config types to include a ranged integer...

ShadowCat117 replied:
I'll use onConfigUpdate for now, but if/when ranged integer gets added it can be changed to that

ryanzhoudev commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,221 @@
+    public void onSlotClicked(ContainerClickEvent e) {
+        if (e.getContainerMenu().containerId != containerId) {
+            return;
+        }
```
```suggestion
        if (e.getContainerMenu().containerId != containerId) return;
```

ryanzhoudev commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,221 @@
+
+        if (currentPage == pageDestination) {
+            return;
+        }
```
```suggestion
        if (currentPage == pageDestination) return;
```

ryanzhoudev commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,221 @@
+
+        // Prevent buying a new page
+        if (e.getSlot().index == NEXT_PAGE_SLOT
+                && e.getSlot().getItem().getHoverName().getString().contains("§c")
```
See if you can use ContainerModel#isLastBankPage

ryanzhoudev commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,221 @@
+
+        StyledText title = StyledText.fromComponent(screen.getTitle());
+
+        Matcher pageMatcher = title.getMatcher(BANK_PAGE_MATCHER);
```
Running this regex on every render is expensive, maybe there is a way to check when the page updates instead?

magicus replied:
Yeah, the render method does way to much. I think we can solve this even without a render method. Can't you just replace the items in the corresponding slots with something that has the wanted icon as texture, and just use vanilla rendering?

ShadowCat117 replied:
Had a go at this, I've got it not running the regex each time but obviously it still calls `renderQuickJumpButton` each time.

But I've not been able to get it to replace the item, I'm not even sure what item uses the texture anyway, it might be NBT data.

If you're able to point me in the right direction I can try again, but no rush if no features aren't being merged until 2.0.3 is released

Mcrtin replied:
try changing the item tag CustomModelData

ryanzhoudev commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,221 @@
+        // When config limits 1-21, change to just customJumpDestinations.get(buttonIndex)
+        int buttonDestination = Math.max(1, Math.min(MAX_BANK_PAGES, customJumpDestinations.get(buttonIndex)));
+
+        buttonSlot.getItem().setHoverName(Component.literal("§7Jump to Page " + buttonDestination));
```
Can you use the ChatFormatting enum instead of raw § codes?

magicus replied:
I think that is fine. We do that elsewhere, I think. @kristofbolyai, what do you say?

kristofbolyai replied:
We use them in a few places yes, while it is usually recommended to use ChatFormatting, it is not strictly necessary.

ryanzhoudev commented:
Thanks for helping port this over!

magicus commented:
Can we please wait with merging feature PRs until after the new 2.0.3 beta is released?

ShadowCat117 committed:
Improvements & cleanup

Commit:
ci: spotless formatting

ShadowCat117 committed:
Don't check regex in onRenderSlot

ShadowCat117 committed:
Use ChatFormatting

Commit:
ci: spotless formatting

magicus commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,226 @@
+@ConfigCategory(Category.INVENTORY)
+public class CustomBankPagesFeature extends Feature {
+    // Change to ranged integer when implemented
+    @RegisterConfig
```
This duplication irks me, but I guess it is up to us not providing a proper array type for configs... It is the same with command hotkeys, iirc.

ShadowCat117 replied:
Yeah, I've been looking into the custom bank page names and that's going to have 21 configs...

magicus commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,226 @@
+
+        if (onLastPage) {
+            quickJumping = false;
+            pageDestination = currentPage;
```
What is pageDestination representing? I don't follow the flow here.

ShadowCat117 replied:
pageDestination is the page you want to jump to, so starts out as 1 so when you open the bank so that it doesn't auto start jumping somewhere.

When you click the next/previous buttons that will increment/decrement the value so that it doesn't auto jump back to the page you were on. This is also needed because the scroll feature can't just tell this feature that a scroll was performed.

Then once you click one of the jump buttons it will use the destination to find which of the default destinations is the closest and each time a new page is rendered, if you're on the last page you own, it will set this page to your destination to stop it from trying to go back to the destination if you hit the previous page button.

Since I don't own page 18, if I tried to jump to 18 it would take me to 17 and then if I clicked the previous button it would try and go back go 18 so destination has to get set to current page if on the last page

magicus replied:
But if you were at the last page and wanted to quickjump to the first page? Wouldn't this cancel that behavior?

ShadowCat117 replied:
Looking at it, yes I think it would.

Changing it to`if (onLastPage && pageDestination > currentPage)` should fix it I think

magicus commented:
I would like to spend some more time on this PR. Please give me a day or two to fully understand if this is okay, or need more work.

ShadowCat117 commented:
> I would like to spend some more time on this PR. Please give me a day or two to fully understand if this is okay, or need more work.

Yeah no worries :)

magicus commented:
I am still struggling to really make heads or tails of the flow in this feature. I think it is unnecessary convoluted. My gut feeling is that we have the wrong state stored in the instance, but I can't pin-point the problem, not give any more suggestions at this point.

However, I also think it will (fully, or at least mostly) work as intended, and it is a welcome addition. 

I realize I cannot give much more advice on the code as part of a review. Instead, I suggest that we merge this as it is now, and then I might give it a shot at trying to refactor this once it is part of the code base. Does that seem ok to you @JamieCallan117?

ShadowCat117 commented:
> I am still struggling to really make heads or tails of the flow in this feature. I think it is unnecessary convoluted. My gut feeling is that we have the wrong state stored in the instance, but I can't pin-point the problem, not give any more suggestions at this point.
> 
> However, I also think it will (fully, or at least mostly) work as intended, and it is a welcome addition. 
> 
> I realize I cannot give much more advice on the code as part of a review. Instead, I suggest that we merge this as it is now, and then I might give it a shot at trying to refactor this once it is part of the code base. Does that seem ok to you @JamieCallan117?

I did have a thought last night on a better way to render the buttons so if you give me a few hours to try that out and I'll see if I can make some improvements elsewhere whilst I'm at it, will also add some comments to hopefully clear some things up. But after that, sure :)

magicus commented:
> I did have a thought last night on a better way to render the buttons so if you give me a few hours to try that out and I'll see if I can make some improvements elsewhere whilst I'm at it, will also add some comments to hopefully clear some things up. But after that, sure :)

I still think the best way would be if we could actually replace the item for the jump slots. But that means we would need to find an item in the Wynn resource pack that seems to fit. I did look through some recently, but found no really good ideas. Best perhaps would be the teleport scroll icon, it looks kind of like you are going to a "page". If we did that. we could skip our individual rendering at all.

ShadowCat117 commented:
> I still think the best way would be if we could actually replace the item for the jump slots. But that means we would need to find an item in the Wynn resource pack that seems to fit. I did look through some recently, but found no really good ideas. Best perhaps would be the teleport scroll icon, it looks kind of like you are going to a "page". If we did that. we could skip our individual rendering at all.

Ok, I've found 3 potential items, I think the scroll with the background would be my choice as it looks more like a button than an item in your bank. The ability tree arrow doesn't really fit the scene but it would be the most obvious as a button.

What do you think?
![2023-07-11_13 42 26](https://github.com/Wynntils/Artemis/assets/8337467/902b87c3-2a39-466b-a322-cc39b9ecab74)

ShadowCat117 commented:
An extra thought actually, could have a config option for any of those choices or no button. Just deciding which one to use as the default then

magicus commented:
Too many config choices is costly as well. I think the scroll with the background looks perfect for the purpose! You could also add the Legacy overlay with the target bank page number, by setting the item count to the target page. That will also give a bit of a clue what page this is taking you to.

ShadowCat117 commented:
> Too many config choices is costly as well. I think the scroll with the background looks perfect for the purpose! You could also add the Legacy overlay with the target bank page number, by setting the item count to the target page. That will also give a bit of a clue what page this is taking you to.

Ooo yes that's a good idea.

I'm almost finished rewriting the flow of it and will push that soon

ShadowCat117 committed:
Bit of refactoring and uses item rendering for buttons instead of custom texture

ShadowCat117 commented:
I think it's a bit better but if you still want to improve it further then feel free too.

Also, setting the button items still happens in `onRenderSlot` I tried doing it in a few other events but it didn't work in the ones I tried, so I might just not know the best one for it.

Mcrtin commented:
did you try the set slot event?

ShadowCat117 commented:
> did you try the set slot event?

I did, but there didn't seem to be a method to set the ItemStack, only get

magicus commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,215 @@
+                buttonFivePage.get(),
+                buttonSixPage.get());
+
+        if (quickJumping && pageDestination > lastPage) {
```
Try converting this based on
`if (!quickJumping) return;`

and then simplify the following if statements

magicus commented:
Overall, this looks much better and closer to what I had in mind!

The "is on last page" detection is still somewhat spread out. I'm not sure if all checks for that are needed, perhaps it is a bit of safety guards as well?

ShadowCat117 committed:
Simplify if statement

magicus commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    private void jumpToDestination() {
+        quickJumping = true;
```
There is a bit of structure here that seems to be longing to be exposed to the world. :-) Basically, this is a glorified switch case based on the value of `currentPage`. Can you help bring that structure forth?

magicus commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,215 @@
+            }
+
+            if (closestDistance < currentDistance) {
+                clickJumpButton(QUICK_JUMP_DESTINATIONS.indexOf(closest));
```
Can you extract this into a tryUsingJumpButtons(), that returns true if successful?

magicus commented:
> I did, but there didn't seem to be a method to set the ItemStack, only get

You need to add this functionality. In `SlotMixin`, do something like this (untested code):
```
        SetSlotEvent.Pre event = new SetSlotEvent.Pre(container, slot, itemStack);
        MixinHelper.post(event);

        ItemStack oldItemStack = container.getItem(slot);
        original.call(container, slot, event.getItemStack());
```

Add a setItemStack() method to the Pre event. (You will need to remove the final designation on the itemStack field in the superclass)

Then you can listen to SetSlotEvent.Pre, and replace the itemstack when it is set.

magicus committed:
Merge branch 'main' into CustomBankQuickJumpButtons

ShadowCat117 committed:
Extract clicking closest jump button to tryUsingJumpButtons

Commit:
ci: spotless formatting

ShadowCat117 committed:
jumpToDestination now uses a switch based on the difference between current and destination page

ShadowCat117 committed:
Merge branch 'CustomBankQuickJumpButtons' of https://github.com/JamieCallan117/Artemis into CustomBankQuickJumpButtons

Commit:
ci: spotless formatting

ShadowCat117 commented:
> > I did, but there didn't seem to be a method to set the ItemStack, only get
> 
> You need to add this functionality. In `SlotMixin`, do something like this (untested code):
> 
> ```
>         SetSlotEvent.Pre event = new SetSlotEvent.Pre(container, slot, itemStack);
>         MixinHelper.post(event);
> 
>         ItemStack oldItemStack = container.getItem(slot);
>         original.call(container, slot, event.getItemStack());
> ```
> 
> Add a setItemStack() method to the Pre event. (You will need to remove the final designation on the itemStack field in the superclass)
> 
> Then you can listen to SetSlotEvent.Pre, and replace the itemstack when it is set.

Not sure I've done this 100% correctly, it is rendering them but it's also rendering them in my inventory in the same slot indexes.

I think it's because `if (!isBankScreen) return;` isn't enough as it's setting the slots for both, but I can't get containerId from the pre event getContainer to check there.

I'll push what I've done so you can see if I've made a mistake.
![2023-07-11_20 53 21](https://github.com/Wynntils/Artemis/assets/8337467/bbebe4b8-451f-4806-841b-9e0fd69b02d0)

ShadowCat117 committed:
Add setItemStack to SetSlotEvent Pre and use for rendering jump buttons

Mcrtin commented:
try ContainerSetSlotEvent and compare the container id like this (in ClientPacketListener#handleContainerSetSlot):
```java
if (packet.getContainerId() == -1) {
			if (!(this.minecraft.screen instanceof CreativeModeInventoryScreen)) {
				player.containerMenu.setCarried(itemStack);
			}
		} else if (packet.getContainerId() == -2) {
			player.getInventory().setItem(i, itemStack);
		} else {
			boolean bl = false;
			Screen var7 = this.minecraft.screen;
			if (var7 instanceof CreativeModeInventoryScreen creativeModeInventoryScreen) {
				bl = !creativeModeInventoryScreen.isInventoryOpen();
			}

			if (packet.getContainerId() == 0 && InventoryMenu.isHotbarSlot(i)) {
				if (!itemStack.isEmpty()) {
					ItemStack itemStack2 = player.inventoryMenu.getSlot(i).getItem();
					if (itemStack2.isEmpty() || itemStack2.getCount() < itemStack.getCount()) {
						itemStack.setPopTime(5);
					}
				}

				player.inventoryMenu.setItem(i, packet.getStateId(), itemStack);
			} else if (packet.getContainerId() == player.containerMenu.containerId && (packet.getContainerId() != 0 || !bl)) {
				player.containerMenu.setItem(i, packet.getStateId(), itemStack);
			}
		}
```

ShadowCat117 commented:
> try ContainerSetSlotEvent and compare the container id like this (in ClientPacketListener#handleContainerSetSlot):

Doesn't look like that's going to work either, can match the container ID's but the actual items don't render, added the same setItemStack method that I added to `SetSlotEvent` but still isn't showing the item

magicus commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,238 @@
+    @SubscribeEvent
+    public void onSetSlot(SetSlotEvent.Pre e) {
+        if (!isBankScreen) return;
+
```
Try something like
```suggestion

if (e.getContainer() instanceof Inventory) return;
```

ShadowCat117 committed:
Return if inventory

ShadowCat117 commented:
Should all work now. If you're happy for this to be merged, shall I make another PR for the same thing for the misc bucket or add it into this one?

magicus commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,240 @@
+        if (!isBankScreen) return;
+
+        if (e.getSlot().index == NEXT_PAGE_SLOT && Models.Container.isLastBankPage(screen)) {
+            lastPage = currentPage;
```
This is all that is left of the rendering. I think we can move it to a more suitable location. onSetSlot should work fine, I think..?

ShadowCat117 replied:
I did try that, but wasn't able to get a reference to the screen to pass in to isLastBankPage. Any idea on how to get it?

magicus replied:
The ContainerModel is a bit wonky designed to work with screens... The best solution would probably be to look for e.g. MenuEvent.MenuOpenedEvent and check getTitle(). This requires rewriting the ContainerModel functionality a bit.

But since this is even more work, I am totally fine with keeping this for now. We can clean this up later as well, if/when we fix the ContainerModel.

ShadowCat117 replied:
Ok, I'll add a comment there as a reminder and then it should be good to go and I'll work on the misc bucket

magicus replied:
I did realize though that a possibly better solution than the SetSlotEvent is to use ContainerSetContentEvent.Pre, where you get all the items at once, and can just replace the individual item stacks. You will not get a container name, but you do get a containerId which you can use to verify that it is the correct container. (But for this to work, you will need to save the containerId, perhaps from a MenuOpenedEvent.)

You have done enough of work on this, so I fully understand if you want to pass on doing this. Let me know if that is the case, and I'll approve it as it is. But if you strive for perfection (and a spot in the Wynntils dev team? :-)), please go ahead!

ShadowCat117 replied:
Did try using ContainerSetContentEvent.Pre yesterday after Mcrtin suggested it and the items didn't render but I'll give it another try quick

magicus replied:
Have a look at how `ItemHandler.onContainerSetContent` works...

ShadowCat117 replied:
I'm definitely close, it renders them once I go to a different page but not the first time the container opens.

ShadowCat117 replied:
This is what I've got, I'm not sure why it doesn't work on the first open, I checked the hovername of the items after the loop by calling e.getItems() again and they did have the new names but when it goes to render it's still using the old items.

```
@SubscribeEvent
    public void onContainerSetContent(ContainerSetContentEvent.Pre e) {
        if (e.getContainerId() != containerId) return;

        List<ItemStack> items = e.getItems();

        for (int buttonSlot : BUTTON_SLOTS) {
            ItemStack jumpButton = new ItemStack(Items.DIAMOND_AXE);
            jumpButton.setDamageValue(92);

            CompoundTag jumpTag = jumpButton.getOrCreateTag();
            jumpTag.putInt("HideFlags", 6);
            jumpTag.putBoolean("Unbreakable", true);
            jumpButton.setTag(jumpTag);

            int buttonIndex = BUTTON_SLOTS.indexOf(buttonSlot);
            int buttonDestination = customJumpDestinations.get(buttonIndex);

            jumpButton.setCount(buttonDestination);

            jumpButton.setHoverName(Component.literal(ChatFormatting.GRAY + "Jump to Page " + buttonDestination));

            items.set(buttonSlot, jumpButton);
        }
    }
```

magicus replied:
Hm, I can't say straight up what's wrong. Can you check by debugging if you even end up in the method for the first page? Is the containerId correct, so you actually go on to change the items, even on the first page?

Do you get SetSlotEvent afterwards (that is how it is usually sent; don't know why)? Can you check if you get that for the first page also? If you do, can you see if you have gotten your new itemstack or if the old one is still in place?

ShadowCat117 replied:
Did do a bit of debugging, I called e.getItems after the loop and checked the hovernames of items at those indexes and it was updated correctly. Then in SetSlot I think it had the old ItemStack so hadn't updated there, but then it does work for every other page which is strange

magicus replied:
This is really strange. I noted recently that ContainerSetContentEvent.Pre is called twice every time you open a new container, and then only once. Maybe this is related somehow... Otherwise I don't know. In worst case, you'll have to drop this change and go back to the per-slot setting.

ShadowCat117 replied:
Yeah I did notice that when the bank first opens it's called once then every other page it's called twice. I wasn't sure if that was something on the servers end or was actually to do with the event. Well if someone figures it out in the future can always come back and retry it but for now guess it'll have to stay in SetSlot

magicus replied:
In the code above, you seem to have a `containerId` field. Is it possible you can use that part of the code, to check if it is the last bank screen, instead of the render function? If we skip the ContainerSetContentEvent part.

ShadowCat117 replied:
I was getting that in onScreenInit, and I don't think the items have rendered by that point so it won't be able to check the name of the item. I can double check though.

Also, just as a quick fyi in case you were about to merge this, I've found a potential issue whilst making this for the misc bucket where if you try to jump to your final page which is 1 below one of the hard coded jump pages, it won't jump anywhere. So I'm just adding a bit more to tryUsingJumpButtons

ShadowCat117 replied:
Actually I'm not sure if the method I was thinking of will work. 

Aside from a render event constantly checking the title of the container to see if it's where we're supposed to be, is there a way to check if after we've tried to click a slot that it has actually moved page?

Also, calling isLastBankPage inside onScreenInit didn't work,

ShadowCat117 replied:
I've fixed this by adding the last pages to storage and in `tryUsingJumpButtons` when it calculates the closest jump, it won't allow it if the jump is greater than the last page.

I've added this method to `ContainerModel` which also eliminates the need for checking for last page in the render event, as this is called in `onScreenInit`.

```
public int updateFinalPage(Screen screen) {
        int currentPage = getCurrentBankPage(screen);

        if (isBankScreen(screen)) {
            if (currentPage > finalBankPage.get()) {
                finalBankPage.store(currentPage);
            }

            return finalBankPage.get();
        } else if (isBookshelfScreen(screen)) {
            if (currentPage > finalBookshelfPage.get()) {
                finalBookshelfPage.store(currentPage);
            }

            return finalBookshelfPage.get();
        } else if (isMiscBucketScreen(screen)) {
            if (currentPage > finalMiscBucketPage.get()) {
                finalMiscBucketPage.store(currentPage);
            }

            return finalMiscBucketPage.get();
        }

        return currentPage;
    }
```

The only problem with this, is that because the default values for final pages is 1, it won't let you quick jump anywhere until you start visiting more pages. I could fix this by having the default values be the maximum number of pages, but then I would need to call `isLastBankPage` again and that would still have to be in the render event.

magicus replied:
Ok, I decided to go to the bottom of this. `isLastBankPage` is just a bit stupidly implemented for our case, since it tries to do too much.

The core thing it does is
```
StyledText.fromComponent(cs.getMenu().getSlot(8).getItem().getHoverName())
                        .endsWith(LAST_BANK_PAGE_STRING);
```
and this you could test from the full container update event.

So my suggestion is that you basically extract this part into a separate method in ContainerModel, like `isItemIndicatingLastBankPage`, and move the 8 to a public constant `LAST_BANK_PAGE_SLOT`. Then call this new function from the old isLastBankPage(), and also use it in your new code.

ShadowCat117 replied:
Should all be done now, I'll just make the changes on my misc bucket branch and check the jumping to last page works there and this should be good to go

ShadowCat117 replied:
Yup, seems to work there so once you visit your final page once it'll all work fine, hopefully most people will figure that out and we won't get too many complaints of people thinking the feature is broken

magicus commented:
I found one more small thing, after all your other changes. :-)

But when that is done, I believe we can merge this.

I suggest we merge this, and then you start a new PR based on main with this PR included, that refactors this code so it supports the misc bucket as well. Better than to start mucking about more with this PR now.

ShadowCat117 committed:
Remove unused variable

magicus committed:
Merge branch 'main' into CustomBankQuickJumpButtons

magicus commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,237 @@
+                if (!tryUsingJumpButtons()) {
+                    if (pageDestination < currentPage && lastPage != currentPage) {
+                        clickPreviousPage();
+                    } else if (lastPage != currentPage) {
```
Could you please invert these checks, so you test "is currentPage ...?"

ShadowCat117 committed:
Store lastBankPage and improve last page checking

ShadowCat117 committed:
Merge branch 'main' into CustomBankQuickJumpButtons

ShadowCat117 committed:
Get final page in onScreenInit

ShadowCat117 committed:
Fix quick jumping directly to final page

magicus commented on common/src/main/java/com/wynntils/features/inventory/CustomBankPagesFeature.java:
```diff
@@ -0,0 +1,237 @@
+
+        switch (pageDifference) {
+            case 1 -> {
+                if (lastPage != currentPage) {
```
```suggestion
                if (currentPage != lastPage) {
```

magicus committed:
Change comparison order on one more place

magicus reviewed (approved):
Finally!!!!!!1!11!!! :-)

Thank you so much for hanging in there! It has been an extraordinary amount of back and forth for what seemed to be a simple feature. But I'm very happy you stayed with us, because this implementation is so much better than the first iteration.

magicus committed:
Merge branch 'main' into CustomBankQuickJumpButtons

ShadowCat117 commented:
No worries! Learnt some more about the project and how to do certain things so future features can hopefully be easier :)
